### PR TITLE
New package: maker

### PR DESCRIPTION
**Pull request by glennpj**
This PR adds the maker package and its dependencies. Note that in addition to the new packages that are in this PR, this PR also depends on the following being merged:

- #12897 
- #12898 
- #12899 
- #12900 
- ~~#12901~~ 
- ~~#12902~~ 
- #12903 
- #12939

### CODE REVIEW

**Comment by glennpj** [comment]
There are a couple of maker perl modules that I missed

**Comment by glennpj** [comment]
This is now ready for review.

File: var/spack/repos/builtin/packages/perl-bioperl/package.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PerlBioperl(PerlPackage):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/perl-bioperl/package.py]: I see a `perl-bio-perl` package. Is this related?

File: var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PerlPerlUnsafeSignals(PerlPackage):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py]: I wouldn't mind if this were called `PerlUnsafeSignals`

File: var/spack/repos/builtin/packages/perl-termreadkey/package.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PerlTermreadkey(PerlPackage):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/perl-termreadkey/package.py]: I see an existing `perl-term-readkey` package - would that be suitable in place of this?

File: var/spack/repos/builtin/packages/perl-bioperl/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: Hmm, I think that may be an older version, where 1.007002 == 1.7.2. I would have to look into this more. It could be that the host repository has changed, along with the version numbering scheme. Even still, based on the tarfile name, I am not sure why there is a '-' in 'bio-perl'.

File: var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py (hunk 2 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py]: I know, but the package is called 'Perl-Unsafe-Signals`. A similar thing occurs with a few R packages and we made those consistent with the upstream name. I am fine either way though.

File: var/spack/repos/builtin/packages/perl-termreadkey/package.py (hunk 3 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-termreadkey/package.py]: That looks like the same package, with this PR having a newer version, but again, why the extra '-'? Should I merge the new version into the existing package, or rename the existing one to not use the extra '-'?

File: var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py (hunk 2 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py]: Oh, were you referring to the package name or just the class name?

File: var/spack/repos/builtin/packages/perl-bioperl/package.py (hunk 1 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: The naming is frequently based off of the camel case in the URL. In this case the `https://cpan.metacpan.org/authors/id/C/CD/CDRAUG/BioPerl-1.7.6.tar.gz` contains `BioPerl-1.7.6.tar.gz` and if you name this class accordingly (`PerlBioPerl`) this is referred to on the command line as `perl-bio-perl` (a dash is inserted between "words" where the capital letter separates words).

A similar application would lead to naming the Termreadkey package `PerlTermReadKey` and it would be referred to as `perl-term-read-key` but I think whoever created it didn't want the extra dash. The consistent approach would be to name the package based on the homepage project name.

There is some leeway in how packages are named (other than that e.g. Perl packages should start with `Perl...`) so these minor inconsistencies don't bother me, as long as we end up with one Spack package per actual package.

File: var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py (hunk 2 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py]: The class name determines the package name (as mentioned in https://github.com/spack/spack/pull/12904#discussion_r328849743).

Comment (scheibelp) [reply var/spack/repos/builtin/packages/perl-perl-unsafe-signals/package.py]: And to be clear I'm OK with the name as is. I just didn't want you to feel that this name was mandatory. It does help find duplicate packages (but as mentioned elsewhere there are minor inconsistencies so I search based on URL components instead typically).

File: var/spack/repos/builtin/packages/perl-bioperl/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: Looking at the `README` and `CHANGES` files it does appear that 1.007002==1.7.2. Those files refer to the version as 1.7.2. So, I will combine my new file into the existing one.

File: var/spack/repos/builtin/packages/perl-termreadkey/package.py (hunk 3 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-termreadkey/package.py]: I will merge the update into the existing file.

File: var/spack/repos/builtin/packages/perl-bioperl/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: >A similar application would lead to naming the Termreadkey package PerlTermReadKey and it would be referred to as perl-term-read-key but I think whoever created it didn't want the extra dash.

Well, I think there is already an extra dash, the package creator left out the second extra dash, if following the camel case rule. In any event, I used `spack list` to search for packages but did not account for extra '-' characters.

Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: For BioPerl, I had a dependency for `perl-uri`. The older package has a dependency for `perl-uri-escape`. However, `URI::Escape` is in the URI package, so in spack, `perl-uri` and `perl-uri-escape` are the same package, but at different versions.

```
class PerlUri(PerlPackage):
    ...
    homepage = "http://search.cpan.org/~ether/URI-1.72/lib/URI.pm"
    url      = "http://search.cpan.org/CPAN/authors/id/E/ET/ETHER/URI-1.72.tar.gz"

    version('1.72', 'cd56d81ed429efaa97e7f3ff08851b48')
```
```
class PerlUriEscape(PerlPackage):
    ...
    homepage = "https://metacpan.org/pod/URI::Escape"
    url      = "https://cpan.metacpan.org/authors/id/E/ET/ETHER/URI-1.71.tar.gz"

    version('1.71', '247c3da29a794f72730e01aa5a715daf')
```
Unless there is a reason not to, I will create a separate PR to remove `perl-uri-escape` and replace it in `depends_on` statements with `perl-uri`.

Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-bioperl/package.py]: I pushed an update for `perl-bio-perl`. Since the version of 1.007002 was

1. really 1.7.2
1. seen by spack as greater than the current 1.7.6
1. and version constraints needed to be specified

I converted version 1.007002 to 1.7.2. Let me know if that was a bad idea. Also, because the configure method for the older version was overridden I added a `@when` decorator. However, for some reason, with that in place, the `@run_after(configure) decorator in `PerlPackage` was not getting picked up. I opted to override the build and install methods as well for the older version.

File: var/spack/repos/builtin/packages/perl-termreadkey/package.py (hunk 3 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/perl-termreadkey/package.py]: This has been committed.

**Comment by glennpj** [comment]
@scheibelp This is ready for another round of review.

Thanks.

**Comment by scheibelp** [comment]
> Also, because the configure method for the older version was overridden I added a @when decorator. However, for some reason, with that in place, the @run_after(configure) decorator in PerlPackage` was not getting picked up. I opted to override the build and install methods as well for the older version.

OK interesting that sounds like a Spack bug, I'll create an issue for it.

**Comment by scheibelp** [comment]
> I'll create an issue for it.

There already was an issue (#12736) so I linked this to it
